src/main/scala/com/fulcrumgenomics/commons/async/AsyncWriterPool.scala

```diff
@@ -0,0 +1,188 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+package com.fulcrumgenomics.commons.async
+
+import java.io.Closeable
+import java.util.concurrent._
+import java.util.concurrent.atomic.{AtomicBoolean, AtomicReference}
+
+import com.fulcrumgenomics.commons.CommonsDef._
+import com.fulcrumgenomics.commons.io.Writer
+
+/**
+  * A shared thread pool for making asynchronous [[Writer]]s without allocating a thread per writer.
+  * The pool is created with a fixed number of threads which are used to handle write operations for
+  * all writers associated to the pool.
+  *
+  * Writers are added to the pool using the [[pool()]] method, which returns a [[PooledWriter]] that
+  * can be written to in place of the original [[Writer]].  [[Writer]]s of different types can be mixed
+  * within the same pool, and each writer can specify either an unbounded queue or a fixed queue of
+  * independent size when being pooled.
+  *
+  * [[Writer]] can be added to the pool at any time (except aftering calling [[close()]] on the pool.
+  * Writers can be closed individually (and removed from the pool) using either their [[PooledWriter.close()]]
+  * method which runs synchronously on the calling thread, or by using [[PooledWriter.closeAsync()]]
+  * which will execute the close asynchronously and return a [[Future]] upon which `get()` must be called
+  * to re-throw any exceptions that were thrown during the close operation.
+  *
+  * The entire pool can be closed be calling [[close()]] on the pool.  This will asynchronously
+  * close all open writers, shutdown the internal thread pool, and rethrow the first exception (if
+  * any) encountered when closing writers.
+  *
+  * @param threads the number of threads to use to write items.
+  */
+class AsyncWriterPool(threads: Int) extends Closeable {
+  require(threads >= 1, "Cannot create a pool with fewer than 1 thread(s).")
+  private val executor   = Executors.newFixedThreadPool(threads)
+  private val writers    = new CopyOnWriteArrayList[PooledWriter[_]]()
+  private val poolClosed = new AtomicBoolean(false)
+  Range(0, threads).foreach(_ => executor.submit(new Drainer))
+
+  /**
+    * Wraps a [[Writer[A]]] in the context of an [[AsyncWriterPool]] to provide asynchronous operations.
+    * Calls to [[write()]] and related methods place items into the queue, and if necessary, schedule
+    * the queue for draining using the [[executor]] in the enclosing [[AsyncWriterPool]].
+    */
```

Comment:
Perhaps some notes for our future selves is appropriate here.  For example, how we spawn a `Drainer` per thread to actively drain the queues while the pool has writers.  Then how there is one final `drain()` when you close a writer.